File: web/src/lib/image-gantt-segments.ts
export type ImageGanttSegment = {
  key: string;
  label: string;
  label_zh: string;
  label_en: string;
  ms: number;
};

export type ImagePhaseTimingsMs = {
  admit_queue_ms?: number;
  upload_queue_ms?: number;
  ps_queue_ms?: number;
  ss_queue_ms?: number;
  account_queue_ms?: number;
  download_queue_ms?: number;
  upload_ms?: number;
  ps_ms?: number;
  ss_ms?: number;
  sse_stream_ms?: number;
  download_ms?: number;
  wall_clock_ms?: number;
};

const SEGMENT_COLORS: Record<string, string> = {
  queue_wait: "bg-stone-300",
  sse_active: "bg-sky-500",
  poll_resolve: "bg-indigo-400",
  download_ms: "bg-emerald-500",
};

export function segmentColor(key: string): string {
  return SEGMENT_COLORS[key] || "bg-stone-400";
}

export function buildImageTaskGanttSegments(
  phase: ImagePhaseTimingsMs | null | undefined,
  opts?: { createdTs?: number; startedTs?: number },
): ImageGanttSegment[] {
  const pt = phase || {};
  let admit = Number(pt.admit_queue_ms || 0);
  let ssQueue = Number(pt.ss_queue_ms || 0);
  let accountQueue = Number(pt.account_queue_ms || 0);
  let sseStream = Number(pt.sse_stream_ms || 0);
  const ssMs = Number(pt.ss_ms || 0);
  const download = Number(pt.download_ms || 0);
  const createdTs = opts?.createdTs;
  const startedTs = opts?.startedTs;

  if (sseStream <= 0 && createdTs && startedTs) {
    const toResolve = Math.max(0, Math.round((startedTs - createdTs) * 1000));
    const knownQueue = admit + ssQueue + accountQueue;
    if (toResolve > knownQueue) {
      sseStream = Math.max(0, Math.min(ssMs, toResolve - knownQueue));
    }
    if (accountQueue <= 0 && toResolve > admit + ssQueue + sseStream) {
      accountQueue = Math.max(0, toResolve - admit - ssQueue - sseStream);
    }
  }

  const queueWait = admit + ssQueue + accountQueue;
  const pollResolve = Math.max(0, ssMs - accountQueue - sseStream);
  const segments: ImageGanttSegment[] = [];
  const push = (key: string, labelZh: string, labelEn: string, ms: number) => {
    const value = Math.max(0, Math.round(ms));
    if (value <= 0) return;
    segments.push({
      key,
      label: `${labelZh} ${labelEn}`,
      label_zh: labelZh,
      label_en: labelEn,
      ms: value,
    });
  };
  push("queue_wait", "排队", "Queue", queueWait);
  push("sse_active", "SSE生图", "SSE", sseStream);
  push("poll_resolve", "轮询", "Poll", pollResolve);
  push("download_ms", "下载", "Download", download);
  return segments;
}

File: utils/image_gantt_segments.py
from __future__ import annotations

from typing import Any


def _segment(key: str, label_zh: str, label_en: str, ms: int) -> dict[str, Any]:
    value = max(0, int(ms or 0))
    return {
        "key": key,
        "label": f"{label_zh} {label_en}",
        "label_zh": label_zh,
        "label_en": label_en,
        "ms": value,
    }


def build_image_task_gantt_segments(
    phase: dict[str, Any] | None,
    *,
    created_ts: float | None = None,
    started_ts: float | None = None,
) -> list[dict[str, Any]]:
    """Build non-overlapping wall-clock segments for image task Gantt charts."""
    pt = phase if isinstance(phase, dict) else {}
    admit = int(pt.get("admit_queue_ms") or 0)
    ss_queue = int(pt.get("ss_queue_ms") or 0)
    account_queue = int(pt.get("account_queue_ms") or 0)
    sse_stream = int(pt.get("sse_stream_ms") or 0)
    ss_ms = int(pt.get("ss_ms") or 0)
    download = int(pt.get("download_ms") or 0)

    if sse_stream <= 0 and created_ts and started_ts:
        to_resolve = max(0, int(round((float(started_ts) - float(created_ts)) * 1000)))
        known_queue = admit + ss_queue + account_queue
        if to_resolve > known_queue:
            sse_stream = max(0, min(ss_ms, to_resolve - known_queue))
        if account_queue <= 0 and to_resolve > admit + ss_queue + sse_stream:
            account_queue = max(0, to_resolve - admit - ss_queue - sse_stream)

    queue_wait = admit + ss_queue + account_queue
    poll_resolve = max(0, ss_ms - account_queue - sse_stream)

    segments: list[dict[str, Any]] = []
    if queue_wait > 0:
        segments.append(_segment("queue_wait", "排队", "Queue", queue_wait))
    if sse_stream > 0:
        segments.append(_segment("sse_active", "SSE生图", "SSE", sse_stream))
    if poll_resolve > 0:
        segments.append(_segment("poll_resolve", "轮询", "Poll", poll_resolve))
    if download > 0:
        segments.append(_segment("download_ms", "下载", "Download", download))
    return segments
